Check if subpath can do motion.

InnerLife0 opened the pull request:
UPD: After long debates we decided to do our own fix, which will not be a backport from master branch, because backported commit contains some duplicate logic.

`cdbpath_create_motion_path` function checks subpath's params Relids, but do it not in place for Bottleneck locus type.  When Path contains index scan with quals it leads to segfault.
This PR moves require_existing_order flag analysis higher, before SingleQE and SegmentGeneral, in the place where it really makes sense.
Additionally it adds missed Insist's in the places where we can get NULL value accidentally, and warning message, so we can find what going wrong even on release version.

Additional regression test included.

The info below is obsolete.
______
`cdbpath_create_motion_path` function doesn't check subpath's params Relids. When Path contains index scan with quals it leads to segfault.
Fix adds a few additional checks of can motion node be generated atop of subpath. Additional regression test included.
Related to issue #[12228](https://github.com/greenplum-db/gpdb/issues/12228). Partial backport of 9cc1da6 (#[10012](https://github.com/greenplum-db/gpdb/issues/10012)).
______
Fix is partial, because we don’t need to first bug, mentioned in commit [message](https://github.com/greenplum-db/gpdb/commit/9cc1da6163d894d638e2455e046dfdbf0763725b). We just don’t have some new functions to fix them (`add_rowid_to_path`). More, regression tests, provided in fix, works well despite of fact orca provides another seems valid plan.
______
There are functions which ignores NULL value returned by `cdbpath_create_motion_path`. `cdbpath_create_motion_path`  can already return NULL in several cases, but PR increases the probability of such event. As returning NULL from `cdbpath_create_motion_path` is not something new and the original commit doesn't contain any additional fixes, we decided to not add NULL value analysis to `cdbpath_create_motion_path` callers, but rather think about PR and mention it [publicly](https://github.com/greenplum-db/gpdb/pull/12287#issuecomment-877009132) at least.

Commit:
Check if subpath can do motion.
'cdbpath_create_motion_path' function doesn't check subpath's params Relids. When Path contains index scan with quals it leads to segfault.
Fix adds a few additional checks of can motion node be generated atop of subpath. Additional regression test included.
Related to issue #12228. Partial backport of 9cc1da6 (#10012).

Commit:
Removing default(fallback) params check.

maksm90 commented on src/test/regress/expected/rpt_joins.out:
```diff
@@ -2153,10 +2153,49 @@ select max(c1) from pg_class left join t_5628 on true;
+create index test_tz_idx on test_tz(tz);
+set optimizer=off;
+set enable_hashjoin=off;
+set enable_seqscan=off;
```
IMO it makes sense to print original optimal plan without motion above inner subpath.

maksm90 replied:
My bad, the line `set enable_seqscan=off` was used to force index scan in inner subplan without calling `analyze` on table. I withdraw my point above.

leskin-in reviewed (dismissed):
Please run `pgindent` on the code changes.

InnerLife0 committed:
Merge branch 'adb-6.x' into ADBDEV-1715-2

Commit:
Fixing PR #12238. Removing unnecessary check for replicated locus type. Adding missed blank lines. Mentioning broken regression test.

(cherry picked from commit 323752317984de22aba579c13768e600dff274e2)

leskin-in reviewed (dismissed):
Run `pgindent` on the modified code from `src/backend/cdb/cdbpath.c`.

InnerLife0 commented:
> Run `pgindent` on the modified code from `src/backend/cdb/cdbpath.c`.

Changed TODO comment to `pgindent` suggestion. The comment above `if` block left as-is, because it's a part of back-port.

darthunix commented on src/backend/cdb/cdbpath.c:
```diff
@@ -365,6 +377,13 @@ cdbpath_create_motion_path(PlannerInfo *root,
+	 * following FIXME from regression tests will be resolved. FIXME: A
+	 * process terminates during execution, see
+	 * https://github.com/greenplum-db/gpdb/issues/10791
+	 */
```
@InnerLife0 can you explain this point to me? You mean that if we get to this position in the code in 6X, he backend crashes because of issue#10791? And we should not check parameters here because of it? Or I didn't get the idea?

InnerLife0 replied:
Placing the `if` condition here as it was done in master, cause a [crash](https://github.com/greenplum-db/gpdb/pull/12238#issuecomment-871810154) of one regression test, which points to issue#10791. This bug is known( unlike master, the next regression test is commented out), but fixing it is not a goal of this issue. So we point we need to do the check, but only after another issue will be fixed.

darthunix replied:
I am still confused with this explanation. An issue [10791](https://github.com/greenplum-db/gpdb/issues/10791) is not about failing on a building plan stage, but rather about execution of the incorrect plan. You fail on a build plan phase, but it can be fixed with a patch like
```
diff --git a/src/backend/cdb/cdbpath.c b/src/backend/cdb/cdbpath.c
index 3a6e8138303..4905cdbf479 100644
--- a/src/backend/cdb/cdbpath.c
+++ b/src/backend/cdb/cdbpath.c
@@ -378,11 +378,10 @@ cdbpath_create_motion_path(PlannerInfo *root,
         */
 
        /*
-        * TODO: Check if subpath require parameters(like in PR #12238) when the
-        * following FIXME from regression tests will be resolved. FIXME: A
-        * process terminates during execution, see
-        * https://github.com/greenplum-db/gpdb/issues/10791
+        * If the subpath requires parameters, we cannot generate Motion atop of it.
         */
+       if (!bms_is_empty(PATH_REQ_OUTER(subpath)))
+               return NULL;
 
        /* Create CdbMotionPath node. */
        pathnode = makeNode(CdbMotionPath);
diff --git a/src/backend/optimizer/util/pathnode.c b/src/backend/optimizer/util/pathnode.c
index 969292b07f5..496f8f7a04c 100644
--- a/src/backend/optimizer/util/pathnode.c
+++ b/src/backend/optimizer/util/pathnode.c
@@ -1862,6 +1862,8 @@ create_unique_path(PlannerInfo *root, RelOptInfo *rel, Path *subpath,
 
                locus = cdbpathlocus_from_exprs(root, uniq_exprs, opfamilies, numsegments);
         subpath = cdbpath_create_motion_path(root, subpath, NIL, false, locus);
+               if (!subpath)
+                       return NULL;
                /*
                 * We probably add agg/sort node above the added motion node, but it is
                 * possible to add an agg/sort node below this motion node also,
```
I think that we should not remove this default case check but rather make it work. Explain fails because new exits returning NULL of the `cdbpath_create_motion_path()` function were added and the other code should learn to handle them. So we should inspect all the calls of `cdbpath_create_motion_path()` and make them work with NULL result (in the draft patch above I have done it in a single place). Good news are that there only a few calls of `cdbpath_create_motion_path()` in GPDB, so it doesn't seem to be something difficult.

InnerLife0 replied:
We've already discussed this problem with team on one of the voice chat sessions. This is not the only place where NULL value returned by `cdbpath_create_motion_path` is ignored. We decided this fix should not be a part of a current PR, but should be [mentioned](https://github.com/greenplum-db/gpdb/pull/12287#issuecomment-877009132) publicly. The only better way we could do is to create a separate PR by ourselves.

darthunix replied:
@InnerLife0 can you write the arguments to leave this PR partially fixed and do the rest of the job in a separate PR instead of doing a full fix? I suppose I have missed that voice chat.

InnerLife0 replied:
Left a comment in PR head.

Stolb27 commented:
> cdbpath_create_motion_path function doesn't check subpath's params Relids.

@InnerLife0, as we discuss yesterday, this statement is incorrect in common. We pass `require_existing_order` argument. This flag is checked if `subpath` for moving contains references to outer relations ([here](https://github.com/arenadata/gpdb/blob/3d6cc3d62b2b800bf72afa0d4a35333f4c543f00/src/backend/optimizer/util/pathnode.c#L3113-L3116), [here](https://github.com/arenadata/gpdb/blob/3d6cc3d62b2b800bf72afa0d4a35333f4c543f00/src/backend/optimizer/util/pathnode.c#L3349-L3350) and [here](https://github.com/arenadata/gpdb/blob/3d6cc3d62b2b800bf72afa0d4a35333f4c543f00/src/backend/optimizer/util/pathnode.c#L3453-L3454)).

`cdbpath_create_motion_path` has next [contract](https://github.com/arenadata/gpdb/blob/3d6cc3d62b2b800bf72afa0d4a35333f4c543f00/src/backend/cdb/cdbpath.c#L82-L85):

```
 *    Else if require_existing_order is true, NULL is returned if the
 *      motion would not preserve an ordering at least as strong as the
 *      specified ordering; also NULL is returned if pathkeys is NIL
 *      meaning the caller is just checking and doesn't want to add motion.
```

and current PR implementation breaks down that. The function must never return NULL if `require_existing_order` is false.
So insist-checks for function calls that **just checking and doesn't want to add motion** (NIL in pathkeys and require_existing_order in false) looks redundant.

We may use `require_existing_order` instead of the current if-condition to save function behavior. This solution eliminates effect to https://github.com/greenplum-db/gpdb/issues/10791. But we need to understand how to process pathkeys for our case.

It seems that we may apply motion to subpath with references to another part of join if pathkeys are provided in the current 6X implementaion.

@maksm90, what do you think?

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -365,6 +377,12 @@ cdbpath_create_motion_path(PlannerInfo *root,
+	 * If the subpath requires parameters, we cannot generate Motion atop of it.
+	 */
+	if (!bms_is_empty(PATH_REQ_OUTER(subpath)))
+		return NULL;
```
@maksm90 I was confused about the type of subpath varaible. It's the argument passed by reference. So in case of condition is true and there is Materialize on top of subpath, we'll remove it anyway after patch.

InnerLife0 commented:
New version of PR is available. Please, re-review.

darthunix commented on src/backend/cdb/cdbpath.c:
```diff
@@ -136,6 +138,22 @@ cdbpath_create_motion_path(PlannerInfo *root,
 			return subpath;
 		}
 
+		/* No motion needed if subpath can run anywhere giving same output. */
```
I suppose there is no need to move `if (CdbPathLocus_IsGeneral(subpath->locus))` code block. It seems that we only need to move up
```c
		/* Fail if caller refuses motion. */
		if (require_existing_order &&
			!pathkeys)
			return NULL;
```

InnerLife0 replied:
Do you think `require_existing_order` is false or `pathkeys` is not NULL, when locus type is General? I suppose, the processing of locus type General is here, because it can be successfully executed even if mentioned condition is true -processing of General locus doesn't apply any motion.

darthunix replied:
I was wrong =)

darthunix commented on src/backend/optimizer/util/pathnode.c:
```diff
@@ -1565,6 +1565,7 @@ set_append_path_locus(PlannerInfo *root, Path *pathnode, RelOptInfo *rel,
 		else
 		{
 			subpath = cdbpath_create_motion_path(root, subpath, subpath->pathkeys, false, targetlocus);
+			Insist(subpath);
```
Why do we insist here that `subpath` should not be NULL? Do we promise somewhere in the code  that `subpath->pathkeys` is NIL? Of cause we can get a segfault in the next line if a new value of `subpath` is NULL, but may be we should handle it in some other way?

InnerLife0 replied:
This is not about `subpath->pathkeys`. Like in other cases, this call pass false to `require_existing_order`, which means it doesn't expect to get NULL output value.
What other ways of handling do you suggest?

darthunix replied:
At the moment we declare another `cdbpath_create_motion_path()` convention:
- `require_existing_order == true` -> we **may** return NULL (not this case)
- `require_existing_order == false && pathkeys == NULL` -> we **always** return NULL (should insist subpath exists)

We have `require_existing_order == false && pathkeys == unknown`, so I think we should not insist but rather fail with a some other error.

InnerLife0 replied:
> * `require_existing_order == false && pathkeys == NULL` -> we **always** return NULL (should insist subpath exists)

Sorry, but I can't understand. We are talking about this condition, right?
```
if (require_existing_order &&
    !pathkeys)
    return NULL;
```
We return NULL if `require_existing_order == true && pathkeys == NULL`, not `require_existing_order == false` as you writing.
It means, getting NULL calling `cdbpath_create_motion_path` with the specified arguments is unexpected behavior, which is handled by `Insist`.

maksm90 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -81,7 +81,9 @@ cdbpath_cost_motion(PlannerInfo *root, CdbMotionPath *motionpath)
  *      motion would not preserve an ordering at least as strong as the
- *      specified ordering; also NULL is returned if pathkeys is NIL
+ *      specified ordering or if path has parameterization relids, 
+ *      which generally may indicate we can't add motion; 
```
Extra spaces were appended to two lines above

InnerLife0 replied:
Resolved

maksm90 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -885,6 +888,11 @@ cdbpath_distkeys_from_preds(PlannerInfo *root,
  * the equijoin predicates between the outer and inner rel.
  * It comes from select_mergejoin_clauses() in joinpath.c.
+ * 
+ * Besides ordering logic, 
```
Extra spaces were appended to two lines above

InnerLife0 replied:
Resolved

maksm90 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -81,7 +81,9 @@ cdbpath_cost_motion(PlannerInfo *root, CdbMotionPath *motionpath)
+ *      specified ordering or if path has parameterization relids,
+ *      which generally may indicate we can't add motion;
+ *      also NULL is returned if pathkeys is NIL
  *      meaning the caller is just checking and doesn't want to add motion.
```
Comments have to follow the common code convention about maximum number of characters in the line (no more than 80) so your line interruptions after punctuation marks violate this.

InnerLife0 replied:
As the main [rule](https://www.postgresql.org/docs/current/source-format.html) is keeping code "readable in an 80-column window" and, imo, the mentioned code is well-readable, I left this lines unchanged. Your opinion will be taken into account in the future.

maksm90 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -885,6 +888,11 @@ cdbpath_distkeys_from_preds(PlannerInfo *root,
  * the equijoin predicates between the outer and inner rel.
  * It comes from select_mergejoin_clauses() in joinpath.c.
+ *
+ * Besides ordering logic,
```
Similarly about line interruption after comma

InnerLife0 replied:
As the main [rule](https://www.postgresql.org/docs/current/source-format.html) is keeping code "readable in an 80-column window" and, imo, the mentioned code is well-readable, I left this lines unchanged. Your opinion will be taken into account in the future.

Stolb27 commented:
@InnerLife0 please rebase this PR on top of `adb-6.x`. I think we need to revert previous solution (https://github.com/arenadata/gpdb/commit/c99892a069919723cb1c8aa27163fd0a3968e02a).

InnerLife0 commented:
Rebased. Commit ebfb060622bd4e0a789427e29dbd58e49ad28b5e left as is, as it was already cherry picked by me to the new PR on the pivotal side. Another commits squashed to one solid solution, which includes removing of our old solution and can be cherry-picked to the mentioned PR.

Commit:
Removing parts backported from 9cc1da6.
'if' conditions backported from 9cc1da6 partially duplicates the logic over require_existing_order flag. Instead of applying additional logic, we can move require_existing_order flag analysis higher, before SingleQE and SegmentGeneral, in the place where it really make sense.

Adding additional comments and warning message, so we can find what going wrong even on release version.
Adding missed Insist's to cdbpath_create_motion_path callers.
